kraky/ws.py

```diff
@@ -129,32 +109,28 @@ async def disconnect(self, connection_name: str = "main") -> None:
         Arguments:
             connection_name: name of the connection you want to disconnect from
         """
-        if self.connections[connection_name] is not None:
-            try:
-                await self.connections[connection_name]["websocket"].close()
-            except socket.gaierror:
-                self.logger.debug("Socket gaia error - disconnecting anyway.")
-            except websockets.exceptions.ConnectionClosedError:
-                self.logger.debug(
-                    "WebSockets connection closed error - disconnecting anyway."
-                )
-            except websockets.exceptions.ConnectionClosedOK:
-                self.logger.debug(
-                    "WebSockets connection closed ok - disconnecting anyway."
-                )
-            except ConnectionResetError:
-                self.logger.debug("Connection reset error - disconnecting anyway.")
-            del self.connections[connection_name]
+        if (
+            connection_name in self.connections and
+            "websocket" in self.connections[connection_name]
+        ):
+            self.logger.debug(
+                "Closing websocket connection '%s'.", connection_name
+            )
+            self.connections[connection_name]["closed"] = True
+            await self.connections[connection_name]["websocket"].close()
 
     async def _send(self, data: dict, connection_name: str = "main") -> None:
-        while connection_name not in self.connections:
+        while not (
+            connection_name in self.connections and
+            "websocket" in self.connections[connection_name]
+        ):
             await asyncio.sleep(0.1)
         try:
             await self.connections[connection_name]["websocket"].send(json.dumps(data))
             await asyncio.sleep(0.1)
         except socket.gaierror:
             self.logger.debug("Socket gaia error - message not sent.")
-        except websockets.exceptions.ConnectionClosedError:
+        except websockets.exceptions.ConnectionClosed as err:
```

Comment:
@DominiqueFuchs Should we print the err variable or just remove it ?

Reply:
Whoops - I originally wanted to do the same consolidation here as in connect and disconnect (catch the generic ConnectionClosed base exception and print the actual type name in the debug log) but then decided to not overdo it and leave other functions untouched, seems I forgot an undo here 😅

As it seems you're fine with this method to make the code a bit leaner I'll push another commit to make it consistent